cpp/src/arrow/util/task_group_test.cc

```diff
@@ -212,6 +212,41 @@ void TestTasksSpawnTasks(std::shared_ptr<TaskGroup> task_group) {
   ASSERT_EQ(count.load(), (1 << (N + 1)) - 1);
 }
 
+// Try to replicate subtle lifetime issues when destroying a TaskGroup
+// where all tasks may not have finished running.
+void StressTaskGroupLifetime(std::function<std::shared_ptr<TaskGroup>()> factory) {
```

Comment:
I find that barrier-debugging is good to trigger such issues:

```
atomic<bool> barrier = false;
{
  TaskGroup group;

  group.Append([&barrier]) { while (!barrier) continue; /* or maybe do something with group? */ return Status::OK() });

  // Triggers group.Finish() to return early before all tasks are completed
  group.Append([]() {return Status::Invalid("oops")});

  group.Finish()
}

// Unlocks thread with group out of scope.
barrier = true;
```

Reply:
I will try it, thank you.

Reply:
Ok, it took some time, but I managed to write a failing test.